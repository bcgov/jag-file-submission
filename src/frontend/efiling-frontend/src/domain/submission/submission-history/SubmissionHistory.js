import React, { useState, useEffect } from "react";
import { Sidecard, Button } from "shared-components";
import { useLocation } from "react-router-dom";
import queryString from "query-string";
import { getSidecardData } from "../../../modules/helpers/sidecardData";
import SubmissionList from "./SubmissionList";
import SafefyCheck from "../../../components/safety-check/SafetyCheck";
import { getSubmissionHistory } from "./SubmissionHistoryService";
import "./SubmissionHistory.scss";
import { Toast } from "../../../components/toast/Toast";

export default function SubmissionHistory() {
  const [showToast, setShowToast] = useState(false);
  const [submissions, setSubmissions] = useState([]);
  const [visibleSubmissions, setVisibleSubmissions] = useState([]);
  const [search, setSearch] = useState("");
  const aboutCsoSidecard = getSidecardData().aboutCso;
  const csoAccountDetailsSidecard = getSidecardData().csoAccountDetails;

  const location = useLocation();
  const queryParams = queryString.parse(location.search);
  const { applicationCode } = queryParams;

  useEffect(() => {
    getSubmissionHistory(applicationCode)
      .then((response) => {
        setSubmissions(response.data);
        setVisibleSubmissions(response.data);
      })
      .catch(() => setShowToast(true));
  }, [applicationCode]);

  function handlePackageSearch() {
    if (search !== "") {
      setVisibleSubmissions(
        submissions.filter((submission) =>
          String(submission.packageNumber).startsWith(String(search))
        )
      );
    } else {
      setVisibleSubmissions(submissions);
    }
  }

  const searchPackagesElement = (
    <>
      <h3>Search Packages</h3>
      <br />
      <span>Package Number</span>
      <br />
      <input
        data-testid="package-search"
        type="number"
        name="search"
        className="package-search"
        onChange={(e) => setSearch(e.target.value)}
      />
      <Button
        data-testid="btn-package-search"
        label="Search"
        styling="bcgov-normal-blue btn search-btn"
        onClick={handlePackageSearch}
      />
    </>
  );

  const submittedPackagesElement = (
    <>
      <div className="submission-list-block">
        <h3>Submitted Packages</h3>
        <div className="submission-list-container">
          <SubmissionList submissions={visibleSubmissions} />
        </div>
      </div>
    </>
  );

  return (
    <div className="ct-submission-history page">
      <div className="content col-md-8">
        <h1>Submission History</h1>
        <br />
        {searchPackagesElement}
        {showToast && (
          <Toast
            content="Something went wrong while trying to retrieve your submissions."
            setShow={setShowToast}
          />
        )}
        {submittedPackagesElement}
        <br />
        <p>
          The View Filing Packages screen lets you view any packages you filed,
          and any packages that you have been given access to by other people in
          your organization. Packages will be available in this list for 1 year
          after the registry processes them.
        </p>
        <SafefyCheck />
      </div>

      <div className="sidecard">
        <Sidecard sideCard={csoAccountDetailsSidecard} />
        <Sidecard sideCard={aboutCsoSidecard} />
      </div>
    </div>
  );
}
