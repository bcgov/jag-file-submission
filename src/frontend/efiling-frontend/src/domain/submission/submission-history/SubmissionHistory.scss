.ct-submission-history {
  /* Chrome, Safari, Edge, Opera */
  input::-webkit-outer-spin-button,
  input::-webkit-inner-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }

  /* Firefox */
  input[type=number] {
    -moz-appearance: textfield;
  }
  .package-search {
    outline: none;
    margin-right: 20px;
    height: 40px;
  }
  .search-btn {
    min-height: 35px;
    height: 40px;
    padding: 0 30px;
  }
  .submission-list-block {
    margin-top: 5%;
  }
  .submission-list-container {
    margin-top: 3px;
    padding: 20px;
    border: 1px solid lightgray;
  }
}
