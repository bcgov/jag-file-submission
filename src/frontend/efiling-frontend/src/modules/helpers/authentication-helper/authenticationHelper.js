const jwt = require("jsonwebtoken");

/**
 * Encodes/signs a payload with JWT into a JSON Web Token string.
 * @param payload Payload to sign, could be an literal, buffer or string
 */
export function encodeJWT(payload) {
  return jwt.sign(payload, "secret");
}

/**
 * Decodes a JWT token, returning the payload, or null if no payload can be found.
 * @param token a JWT token string to decode
 */
export function decodeJWT(token) {
  const decoded = jwt.decode(token, { complete: true });

  if (decoded && decoded.payload) return decoded.payload;
  return null;
}

export function getJWTData() {
  const token = localStorage.getItem("jwt");

  // get the decoded payload and header
  return decodeJWT(token);
}

export function generateJWTToken(payload) {
  return jwt.sign(payload, "secret");
}

/**
 * Returns the identity provider of the current user.
 */
export function getIdentityProviderAlias() {
  const token = getJWTData();
  return token.identityProviderAlias;
}
