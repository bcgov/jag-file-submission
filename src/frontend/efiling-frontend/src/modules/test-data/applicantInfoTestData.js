const applicantInfo = {
  bceid: "bobross42",
  firstName: "Bob",
  lastName: "Ross",
  email: "bob.ross@example.com",
};

export function getApplicantInfo() {
  return applicantInfo;
}
