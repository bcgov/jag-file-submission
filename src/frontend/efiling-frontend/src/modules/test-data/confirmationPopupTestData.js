/* eslint-disable react/jsx-one-expression-per-line */
import React from "react";

const onButtonClick = () => {};

const mainButton = {
  label: "Cancel",
  styling: "bcgov-normal-white btn",
  onClick: onButtonClick,
};

const confirmButton = {
  label: "Yes, cancel",
  styling: "bcgov-normal-blue btn bcgov-consistent-width",
  onClick: onButtonClick,
};

const cancelButton = {
  label: "No, dont cancel",
  styling: "bcgov-normal-white btn bcgov-consistent-width",
  onClick: onButtonClick,
};

const body = () => (
  <>
    <p>Your files will not be submitted.</p>
    <p>
      You will be returned to:
      <br />
      <b>Original</b> website
    </p>
  </>
);

export function getTestData() {
  const modal = {
    show: false,
    title: "title",
    body,
  };

  return {
    modal,
    mainButton,
    confirmButton,
    cancelButton,
  };
}
