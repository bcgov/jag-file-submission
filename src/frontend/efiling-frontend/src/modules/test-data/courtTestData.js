const court = {
  locationDescription: "Court location",
  fileNumber: "Court file number",
  levelDescription: "Level",
  classDescription: "Class",
  level: "P",
  courtClass: "F",
  location: "Kelowna Law Courts",
};

export function getCourtData() {
  return court;
}
