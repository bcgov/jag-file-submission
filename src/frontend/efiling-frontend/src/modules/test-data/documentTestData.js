const documents = [
  {
    documentProperties: {
      name: "file name 1",
      type: "file type",
    },
    description: "file description 1",
    statutoryFeeAmount: 40,
    isAmendment: null,
    isSupremeCourtScheduling: null,
  },
  {
    documentProperties: {
      name: "file name 2",
      type: "file type",
    },
    description: "file description 2",
    statutoryFeeAmount: 0,
    isAmendment: false,
    isSupremeCourtScheduling: true,
  },
];

export function getDocumentsData() {
  return documents;
}
