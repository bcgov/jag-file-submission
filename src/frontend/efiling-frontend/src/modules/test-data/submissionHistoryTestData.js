export const submissions = [
  {
    packageNumber: 1234567,
    submittedDate: "04/30/1996",
    status: {
      description: "Approved",
    },
    court: {
      location: "Kelowna",
    },
  },
  {
    packageNumber: 7654321,
    submittedDate: "04/30/1996",
    status: {
      description: "Rejected",
    },
    court: {
      location: "Victoria",
    },
  },
];
