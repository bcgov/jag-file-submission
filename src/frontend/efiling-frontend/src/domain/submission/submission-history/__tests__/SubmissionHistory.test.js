import React from "react";
import axios from "axios";
import { render, waitFor, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import MockAdapter from "axios-mock-adapter";
import SubmissionHistory from "../SubmissionHistory";
import { submissions } from "../../../../modules/test-data/submissionHistoryTestData";
import { generateJWTToken } from "../../../../modules/helpers/authentication-helper/authenticationHelper";

const routerDom = require("react-router-dom");

jest.mock("react-router-dom");

describe("Submission History Component", () => {
  let mock;
  beforeEach(() => {
    routerDom.useLocation = jest
      .fn()
      .mockReturnValue({ search: "?applicationCode=FLA" });

    const token = generateJWTToken({
      preferred_username: "username@bceid",
      email: "username@example.com",
      identityProviderAlias: "bcsc",
    });
    localStorage.setItem("jwt", token);

    mock = new MockAdapter(axios);
    window.open = jest.fn();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test("Matches the snapshot", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { asFragment, getByText } = render(<SubmissionHistory />);
    await waitFor(() => getByText("1234567"));

    expect(asFragment()).toMatchSnapshot();
  });

  test("Searching package # updates ui with new list", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { getByTestId, getByText, queryByText } = render(
      <SubmissionHistory />
    );
    await waitFor(() => getByText("1234567"));

    const searchBtn = getByText("Search");
    const searchBar = getByTestId("package-search");

    userEvent.type(searchBar, "123");
    await waitFor(() => {});

    fireEvent.click(searchBtn);
    await waitFor(() => {});

    expect(queryByText("7654321")).not.toBeInTheDocument();
  });

  test("Resetting search value resets visible submissions", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { getByTestId, getByText, queryByText } = render(
      <SubmissionHistory />
    );
    await waitFor(() => getByText("1234567"));

    const searchBtn = getByText("Search");
    const searchBar = getByTestId("package-search");

    userEvent.type(searchBar, "");
    await waitFor(() => {});

    fireEvent.click(searchBtn);
    await waitFor(() => {});

    expect(queryByText("1234567")).toBeInTheDocument();
    expect(queryByText("7654321")).toBeInTheDocument();
  });

  test("Package number redirects to package url - click", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { getByText } = render(<SubmissionHistory />);
    await waitFor(() => getByText("1234567"));

    const packageNumber = getByText("1234567");
    fireEvent.click(packageNumber);
    await waitFor(() => {});

    expect(window.open).toHaveBeenCalled();
  });

  test("Package number redirects to package url - keydown", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { getByText } = render(<SubmissionHistory />);
    await waitFor(() => getByText("1234567"));

    const packageNumber = getByText("1234567");

    fireEvent.keyDown(packageNumber, {
      key: "Enter",
      keyCode: "13",
    });
    await waitFor(() => {});

    expect(window.open).toHaveBeenCalled();
  });

  test("Package number redirects to package url - incorrect input", async () => {
    mock.onGet("/filingpackages").reply(200, submissions);

    const { getByText } = render(<SubmissionHistory />);
    await waitFor(() => getByText("1234567"));

    const packageNumber = getByText("1234567");

    fireEvent.keyDown(packageNumber, {
      key: "Enter",
      keyCode: "9",
    });
    await waitFor(() => {});

    expect(window.open).not.toHaveBeenCalled();
  });
});
