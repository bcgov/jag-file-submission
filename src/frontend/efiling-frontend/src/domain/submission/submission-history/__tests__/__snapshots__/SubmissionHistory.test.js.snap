// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Submission History Component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ct-submission-history page"
  >
    <div
      class="content col-md-8"
    >
      <h1>
        Submission History
      </h1>
      <br />
      <h3>
        Search Packages
      </h3>
      <br />
      <span>
        Package Number
      </span>
      <br />
      <input
        class="package-search"
        data-testid="package-search"
        name="search"
        type="number"
      />
      <button
        class="bcgov-button bcgov-normal-blue btn search-btn"
        data-testid=""
        type="button"
      >
        Search
      </button>
      <div
        class="submission-list-block"
      >
        <h3>
          Submitted Packages
        </h3>
        <div
          class="submission-list-container"
        >
          <div
            class="ct-submission-list"
          >
            <div
              class="table-responsive"
            >
              <table
                class="table"
              >
                <thead>
                  <tr>
                    <th
                      scope="col"
                    >
                      Package #
                    </th>
                    <th
                      scope="col"
                    >
                      Date Submitted
                    </th>
                    <th
                      scope="col"
                    >
                      Application
                    </th>
                    <th
                      scope="col"
                    >
                      Court Location
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    class="table-row"
                  >
                    <td
                      class="border-left"
                    >
                      <span
                        class="file-href"
                        role="button"
                        tabindex="0"
                      >
                        1234567
                      </span>
                    </td>
                    <td>
                      30-Apr-1996
                    </td>
                    <td>
                      xxxxx
                    </td>
                    <td
                      class="border-right"
                    >
                      Kelowna
                    </td>
                  </tr>
                  <tr
                    class="table-row"
                  >
                    <td
                      class="border-left"
                    >
                      <span
                        class="file-href"
                        role="button"
                        tabindex="0"
                      >
                        7654321
                      </span>
                    </td>
                    <td>
                      30-Apr-1996
                    </td>
                    <td>
                      xxxxx
                    </td>
                    <td
                      class="border-right"
                    >
                      Victoria
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
      </div>
      <br />
      <p>
        The View Filing Packages screen lets you view any packages you filed, and any packages that you have been given access to by other people in your organization. Packages will be available in this list for 1 year after the registry processes them.
      </p>
    </div>
    <div
      class="sidecard"
    >
      <div
        class="bcgov-wide-dashboard-spacing"
        style="position: relative;"
      >
        <div
          class="col-12 pt-2"
          style="position: relative;"
        >
          <p />
          <section
            class="bcgov-bluegrey-container"
            id="csoAccountDetailsCard"
          >
            <div
              class="bcgov-container-background bcgov-bluegrey-heading"
            >
              <h2
                class="bcgov-heading-style"
              >
                <div
                  class="bcgov-side-card-row"
                >
                  <div
                    class="bcgov-round-icon-wrapper"
                  >
                    <svg
                      class="bcgov-side-card-icon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                      />
                    </svg>
                  </div>
                  <div
                    class="bcgov-side-card-title"
                  >
                    Your CSO Account
                  </div>
                </div>
              </h2>
            </div>
            <div
              class="bcgov-bluegrey-content"
            >
              <span
                class="bcgov-content-style"
              >
                <p>
                  CSO account 
                  <strong />
                   is linked to your Basic BCeID account 
                  <strong>
                    username
                  </strong>
                   and will be used to file documents. 
                  <a
                    href="undefined/accounts/editProfile.do"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    View your CSO account details
                  </a>
                  .
                </p>
              </span>
            </div>
          </section>
        </div>
      </div>
      <div
        class="bcgov-wide-dashboard-spacing"
        style="position: relative;"
      >
        <div
          class="col-12 pt-2"
          style="position: relative;"
        >
          <p />
          <section
            class="bcgov-bluegrey-container"
            id="aboutCsoCard"
          >
            <div
              class="bcgov-container-background bcgov-bluegrey-heading"
            >
              <h2
                class="bcgov-heading-style"
              >
                <div
                  class="bcgov-side-card-row"
                >
                  <div
                    class="bcgov-round-icon-wrapper"
                  >
                    <svg
                      class="bcgov-side-card-icon"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M11 17h2v-6h-2v6zm1-15C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zM11 9h2V7h-2v2z"
                      />
                    </svg>
                  </div>
                  <div
                    class="bcgov-side-card-title"
                  >
                    About E-File Submission
                  </div>
                </div>
              </h2>
            </div>
            <div
              class="bcgov-bluegrey-content"
            >
              <span
                class="bcgov-content-style"
              >
                <p>
                  E-File Submission is a service to help you securely and electronically file documents with the Courts of British Columbia via Court Services Online. 
                  <a
                    href="undefined/about/index.do"
                    rel="noopener noreferrer"
                    target="_blank"
                  >
                    Learn more about CSO
                  </a>
                  .
                </p>
              </span>
            </div>
          </section>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
